### PR TITLE
feat(overlay): add connected position strategy

### PR DESCRIPTION
**Pull request by jelbourn**
R: @hansl @robertmesserle @kara 

I'm still missing tests for `ViewportRuler`, but I wanted to get the review started while I work on those. 
(and also see if this passes on CI)

### CODE REVIEW

File: src/core/overlay/position/connected-position-strategy.ts [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+
+  /** Adds a preferred position to the end of the ordered preferred position list. */
+  addPreferredPosition(pos: RelativePosition): void {
+    this._preferredPositions.push(pos);
+  }
+
+
+  /**
+   * Gets the horizontal (x) "start"" dimension based on whether the overlay is in an RTL context.
```
Comment (kara) [inline src/core/overlay/position/connected-position-strategy.ts]: Extra `"`

Comment (jelbourn) [inline src/core/overlay/position/connected-position-strategy.ts]: Fixed.

**Comment by jelbourn** [comment]
Added tests for `ViewportRuler`.

**Comment by jelbourn** [comment]
Tests are green now. Swapped the browser setup to use iOS8 instead of iOS7

File: src/core/overlay/position/connected-position-strategy.ts [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+  private _getOriginConnectionPoint(originRect: ClientRect, pos: RelativePosition): Point {
+    const originStartX = this._getStartX(originRect);
+    const originEndX = this._getEndX(originRect);
+
+    let x: number;
+    if (pos.originX == 'center') {
+      x = originStartX + (originRect.width / 2);
+    } else {
+      x = pos.originX == 'start' ? originStartX : originEndX;
+    }
```
Comment (kara) [inline src/core/overlay/position/connected-position-strategy.ts]: Nitpick: You could split this into `_getOriginConnectionX` and `_getOriginConnectionY` methods to make it a bit more focused.

File: src/core/overlay/position/viewport-ruler.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+@Injectable()
+export class ViewportRuler {
+  // TODO(jelbourn): cache the document's bounding rect and only update it when the window
+  // is resized (debounced).
+
+
+  /** Gets a ClientRect for the viewport's bounds. */
+  getViewportRect(): ClientRect {
+    // Use the document element's bounding rect rather than the window scroll properties
+    // (e.g. pageYOffset, scrollY) due to in issue in Chrome and IE where window scroll
```
Comment (kara) [inline src/core/overlay/position/viewport-ruler.ts]: Typo - `in` -> `an`

File: src/core/style/apply-transform.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+ * @param element
+ * @param transformValue
+ */
+export function applyCssTransform(element: HTMLElement, transformValue: string) {
+  // It's important to trim the result, because the browser will ignore the set operation
+  // if the string contains only whitespace.
+  let value = transformValue.trim();
+
+  element.style.transform = value;
+  element.style.webkitTransform = value;
```
Comment (kara) [inline src/core/style/apply-transform.ts]: Do we need anything for moz here or is that covered by `transform`?

File: src/core/overlay/position/connected-position-strategy.ts [hunk 5]
```diff
@@ -0,0 +1,204 @@
+import {PositionStrategy} from './position-strategy';
+import {ElementRef} from 'angular2/core';
+import {ViewportRuler} from './viewport-ruler';
+import {applyCssTransform} from '../../style/apply-transform';
+
+
+
+/**
+ * A strategy for positioning overlays. Using this strategy, an overlay is given an
+ * implict position relative some origin element. The relative position is defined in terms of
```
Comment (kara) [inline src/core/overlay/position/connected-position-strategy.ts]: Typo: `implict` --> `implicit`

**Comment by kara** [comment]
Done reviewing. Just a few typos and questions, but looks good to me. @hansl, @robertmesserle ?

File: src/core/overlay/overlay.ts [hunk 6]
```diff
@@ -41,7 +41,8 @@ export class Overlay {
   constructor(
       @Inject(OVERLAY_CONTAINER_TOKEN) private _overlayContainerElement: HTMLElement,
       private _dynamicComponentLoader: DynamicComponentLoader,
-      private _appViewManager: AppViewManager) {
+      private _appViewManager: AppViewManager,
+      private _positionBuilder: OverlayPositionBuilder) {
   }
```
Comment (robertmesserle) [inline src/core/overlay/overlay.ts]: Single line for `{ }` for consistency?

File: src/core/overlay/position/connected-position-strategy.ts [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+ * a point on the origin element that is connected to a point on the overlay element. For example,
+ * a basic dropdown is connecting the bottom-left corner of the origin to the top-left corner
+ * of the overlay.
+ */
+export class ConnectedPositionStrategy implements PositionStrategy {
+  // TODO(jelbourn): set RTL to the actual value from the app.
+  /** Whether the we're dealing with an RTL context */
```
Comment (robertmesserle) [inline src/core/overlay/position/connected-position-strategy.ts]: - the

**Pull request merged**